core/modules/layout/layout.context.admin.inc

```diff
@@ -0,0 +1,230 @@
+<?php
+module_load_include('inc', 'layout', 'layout.admin');
+
+
+/**
+ * Submit handler for layout_settings_form() that lets the user add a context.
+ */
+function layout_settings_form_context_add($form, &$form_state) {
+  // Remove destination flags when adding contexts, otherwise the user will
+  // be taken to the destination instead of adding a context. This only
+  // affects non-JS behavior.
+  if (isset($_GET['destination'])) {
+    unset($_GET['destination']);
+  }
+  $form_state['redirect'] = 'admin/structure/layouts/manage/' . $form_state['layout']->name . '/context/add';
+}
+
+/**
+ * Form callback; Displays form for adding new contexts to a layout.
+ *
+ * @param Layout $layout
+ *   The layout to which a context is being added.
+ * @param int|null $context_id
+ *   The ID of the context being configured. If adding a new
+ *   context, no value will be passed in.
+ *
+ * @ingroup forms
+ */
+function layout_context_add_form($form, &$form_state, Layout $layout = NULL, $context_id = NULL) {
+  form_load_include($form_state, 'inc', 'layout', 'layout.admin');
+
+  $is_new_context = TRUE;
+  if (isset($context_id)) {
+    $is_new_context = FALSE;
+  }
+
+  $form_state['layout'] = $layout;
+  $form_state['context_id'] = $context_id;
+
+  $form['help'] = array(
+    '#type' => 'help',
+    '#markup' => 'help',
+  );
+
+  $handler = isset($layout->contexts[$context_id]) ? $layout->contexts[$context_id] : NULL;
+
+  $all_context_info = _layout_get_all_info('layout_context');
+  backdrop_sort($all_context_info, array('title'));
+  $form['context'] = array(
+    '#type' => 'select',
+    '#title' => t('Contexts'),
+    '#options' => array(),
+    '#required' => TRUE,
+    '#default_value' => isset($handler->plugin) ? $handler->plugin : NULL,
+    '#parents' => array('context'),
+    '#ajax' => array(
+      'wrapper' => 'context_settings',
+      'callback' => 'layout_context_ajax_style',
+    ),
+  );
+  foreach ($all_context_info as $context_info) {
+    $form['context']['#options'][$context_info['name']] = $context_info['title'];
+  }
```

Comment:
As a non experienced developer, I'd expect this loop to come before the select element, setting the values of say a `$context_options` array variable, and then use that array instead of the empty array in `'#options' => array(),` earlier.

Reply:
This is necessary. If we try to build the list first, then this part `$form['context']['#options']...` would be undefined. We have to make that form element first.